File: day001_hash_tables_base/related_contains_duplicate/solution.py
from typing import List

class Solution:

    # 1. Brute Force Approach: Time O(n^2), Space O(1)
    # Compares every pair of elements.
    def containsDuplicate_brute(self, nums: List[int]) -> bool:
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] == nums[j]:
                    return True
        return False

    # 2. Optimized Approach: Time O(n), Space O(n) - Set
    # Uses a set for O(1) average time complexity lookups.
    def containsDuplicate_optimized(self, nums: List[int]) -> bool:
        seen = set()
        for num in nums:
            if num in seen:
                return True
            seen.add(num)
        return False
    
    # Pythonic One-Liner (O(n) Time/Space)
    def containsDuplicate_pythonic(self, nums: List[int]) -> bool:
        return len(nums) != len(set(nums))

    # Default entry point (using the optimized method)
    def containsDuplicate(self, nums: List[int]) -> bool:
        return self.containsDuplicate_optimized(nums)

File: day001_hash_tables_base/related_contains_duplicate/solution.cpp
#include <vector>
#include <unordered_set>
#include <algorithm>

class Solution {
public:
    // 1. Brute Force Approach: Time O(n^2), Space O(1)
    // Compares every pair of elements.
    bool containsDuplicate_brute(std::vector<int>& nums) {
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            for (int j = i + 1; j < n; ++j) {
                if (nums[i] == nums[j]) {
                    return true;
                }
            }
        }
        return false;
    }

    // 2. Optimized Approach: Time O(n), Space O(n) - Hash Set
    // Uses a set for O(1) average time complexity lookups.
    bool containsDuplicate_optimized(std::vector<int>& nums) {
        std::unordered_set<int> seen;
        for (int num : nums) {
            // Check if the number has already been added (duplicate found)
            if (seen.count(num)) {
                return true;
            }
            // Add the current number to the set
            seen.insert(num);
        }
        return false;
    }
    
    // Default entry point (using the optimized method)
    bool containsDuplicate(std::vector<int>& nums) {
        return containsDuplicate_optimized(nums);
    }
};